pkgs/development/interpreters/python/python-packages-base.nix

```diff
@@ -14,26 +14,40 @@ let
 
   # Derivations built with `buildPythonPackage` can already be overridden with `override`, `overrideAttrs`, and `overrideDerivation`.
   # This function introduces `overridePythonAttrs` and it overrides the call to `buildPythonPackage`.
+  #
+  # Overridings specified through `overridePythonAttrs` will always be applied
+  # before those specified by `overrideAttrs`, even if invoked after them.
   makeOverridablePythonPackage =
     f:
-    lib.mirrorFunctionArgs f (
+    let
+      mirrorArgs = lib.mirrorFunctionArgs f;
+    in
+    mirrorArgs (
       origArgs:
       let
         args = lib.fix (
           lib.extends (_: previousAttrs: {
             passthru = (previousAttrs.passthru or { }) // {
-              overridePythonAttrs = newArgs: makeOverridablePythonPackage f (overrideWith newArgs);
+              inherit overridePythonAttrs;
             };
           }) (_: origArgs)
         );
         result = f args;
-        overrideWith = newArgs: args // (if pkgs.lib.isFunction newArgs then newArgs args else newArgs);
+
+        overrideWith = newArgs: args // lib.toFunction newArgs args;
```

Comment:
Slightly off topic:

`overrideAttrs` supports being called with:
1. `attrs`
2. `(prev: attrs)`
3. `(final: prev: attrs)`

At some point, it'd be nice if `overridePythonAttrs` also supported the third style (overlay-style extension function), currently it only supports the first two.

I assume it'd be possible to encapsulate that within `overrideWith`? In another PR, of course.

Reply:
It can be done in `overrideWith`, but I don't feel like letting `overridePythonAttrs` take fixed-point arguments.
1. `overridePythonAttrs` is a custom overrider and `buildPythonPackage` argument mapping contains a lot of historical burden (e.g., `checkPhase` -> `installCheckPhase`). I'm leaning toward deprecating `buildPythonPackage` after fixing the `finalAttrs` reference for `<pkg>.overrideAttrs` and fixed-point arguments.
2. We need to subtract the `stdenv` argument as implemented in #271762, and supporting fixed-point arguments make it much more challenging (if not impossible to implement).